From 71422667cba5ba7249d9fbaf855caeba133768ca Mon Sep 17 00:00:00 2001
From: David Redondo <qt@david-redondo.de>
Date: Tue, 29 Jul 2025 09:31:01 +0200
Subject: [PATCH 329/553] Handle subsurfaces with scaling correctly

We need to use device coordinates to copy
the correct region, not logical window size.
Also paint the subsurface with scaling and
not scale down.
Use the passed offset instead of window
position, sometimes they dont match.

Change-Id: I2019dd093bdfb6eb225845606509820523f10034
Reviewed-by: David Redondo <qt@david-redondo.de>
(cherry picked from commit 1f67152e21c5a9527a1d4b046ee0fd521c0fb808)
---
 src/plugins/platforms/wayland/qwaylandshmbackingstore.cpp | 8 +++++---
 1 file changed, 5 insertions(+), 3 deletions(-)

diff --git a/src/plugins/platforms/wayland/qwaylandshmbackingstore.cpp b/src/plugins/platforms/wayland/qwaylandshmbackingstore.cpp
index d8fc7b18ca3..9a67190595c 100644
--- a/src/plugins/platforms/wayland/qwaylandshmbackingstore.cpp
+++ b/src/plugins/platforms/wayland/qwaylandshmbackingstore.cpp
@@ -259,7 +259,6 @@ bool QWaylandShmBackingStore::scroll(const QRegion &region, int dx, int dy)
 
 void QWaylandShmBackingStore::flush(QWindow *window, const QRegion &region, const QPoint &offset)
 {
-    Q_UNUSED(offset)
     // Invoked when the window is of type RasterSurface or when the window is
     // RasterGLSurface and there are no child widgets requiring OpenGL composition.
 
@@ -268,9 +267,12 @@ void QWaylandShmBackingStore::flush(QWindow *window, const QRegion &region, cons
     // however so no need to reimplement that.
     if (window != this->window()) {
         auto waylandWindow = static_cast<QWaylandWindow *>(window->handle());
-        auto newBuffer = new QWaylandShmBuffer(mDisplay, window->size(), mBackBuffer->image()->format(), mBackBuffer->scale(), mEventQueue);
+        const auto scale = waylandWindow->scale();
+        auto newBuffer = new QWaylandShmBuffer(
+                mDisplay, window->size() * scale, mBackBuffer->image()->format(),
+                mBackBuffer->image()->devicePixelRatio(), mEventQueue);
         newBuffer->setDeleteOnRelease(true);
-        QRect sourceRect(window->position(), window->size());
+        QRect sourceRect(offset * scale, window->size() * scale);
         QPainter painter(newBuffer->image());
         painter.drawImage(QPoint(0, 0), *mBackBuffer->image(), sourceRect);
         waylandWindow->safeCommit(newBuffer, region);
-- 
2.51.2

